packages/proto-signing/src/codec/cosmos/bank/v1beta1/tx.ts

```diff
@@ -0,0 +1,324 @@
+/* eslint-disable */
+import { Coin } from "../../../cosmos/base/v1beta1/coin";
+import { Input, Output } from "../../../cosmos/bank/v1beta1/bank";
+import { Reader, Writer } from "protobufjs/minimal";
+
+/**
+ *  MsgSend represents a message to send coins from one account to another.
+ */
+export interface MsgSend {
+  fromAddress: string;
+  toAddress: string;
+  amount: Coin[];
+}
+
+/**
+ *  MsgSendResponse defines the Msg/Send response type.
+ */
+export interface MsgSendResponse {}
+
+/**
+ *  MsgMultiSend represents an arbitrary multi-in, multi-out send message.
+ */
+export interface MsgMultiSend {
+  inputs: Input[];
+  outputs: Output[];
+}
+
+/**
+ *  MsgMultiSendResponse defines the Msg/MultiSend response type.
+ */
+export interface MsgMultiSendResponse {}
+
+const baseMsgSend: object = {
+  fromAddress: "",
+  toAddress: "",
+};
+
+const baseMsgSendResponse: object = {};
+
+const baseMsgMultiSend: object = {};
+
+const baseMsgMultiSendResponse: object = {};
+
+/**
+ *  Msg defines the bank Msg service.
+ */
+export interface Msg {
+  /**
+   *  Send defines a method for sending coins from one account to another account.
+   */
+  Send(request: MsgSend): Promise<MsgSendResponse>;
```

Comment:
In general, it is my belief we should maintain a fork of ts-proto, so that we have flexibility on the exact output we want.

For example, let's take this auto-generated interface in x/bank (related to ADR-031). It doesn't make sense to me to return a promise here. Also, there should be a better integration with `SigningStargateClient` for tx building.



I started exploring this by forking ts-proto and playing around. A good UX example for me would be:

```ts
const signingClient = new SigningStargateClient(...);
const bankMsgClient = new MsgClientImpl(signingClient); // MsgClientImpl is imported from this file

// signingClient (or a wrapper around it) could hold an internal array of `Msg`s, and the following line would add one `Msg` to the array
bankMsgClient.Send(request1, (response) => { console.log(response); }); // Everything will be well-typed.
// We could add more `Msg`s to the tx, by importing `Msg` clients from other modules
stakingMsgClient.Delegate(request2, (response) => { console.log(response); });

// ...

signingClient.signAndBroadcast(...); // Signs and broadcasts tx with all internal `Msg`s.
```

For those interested, I forked ts-proto [here](https://github.com/amaurymartiny/ts-proto/pull/1/files) and it generated [this code](https://github.com/regen-network/regen-js/pull/14/files#diff-0b3e07c937587ad311e5fd37408afe77dcf3ba661f5d3c24851ee3282d1d191bR76-R82)

Reply:
> It doesn't make sense to me to return a promise here.

Why? What would be the alternative?

Reply:
Mostly because of multi-Msgs txs. For those txs, this method should add a `x/bank Send Msg` to the tx building client (here `SigningStargateClient`, or a wrapper of it), and that's a sync operation. The tx building client could then do something else, like adding other `Msg`s, or fees/memo.

The signature would look like:

```ts
Send(request: MsgSend, callback?: (response: MsgSendResponse) => void): void;
```

And see my message above for an example usage.

Reply:
If it is a sync operation, then it is clearly not an implementation of the `service`.

But what you outlined there can easily be done with the Promise interface as well. Promises can take as long as you want until they resolve.

```
const signingClient = new SigningStargateClient(...);
const bankMsgClient = new MsgClientImpl(signingClient); // MsgClientImpl is imported from this file

// signingClient (or a wrapper around it) could hold an internal array of `Msg`s, and the following line would add one `Msg` to the array
const pendingSend = bankMsgClient.Send(request1); // Everything will be well-typed.
// We could add more `Msg`s to the tx, by importing `Msg` clients from other modules
const pendingDelegate = stakingMsgClient.Delegate(request2);

// ...

signingClient.signAndBroadcast(...); // Signs and broadcasts tx with all internal `Msg`s.

// bankMsgClient and stakingMsgClient observe chain and wait for execution

const [response1, response2] = await = Promise.all([pendingSend, pendingDelegate]);
```

Reply:
That would make sense too! It also seemed to me like the obvious way to do it with ts-proto, however, I find it slightly less idiomatic:

```ts
const pendingSend = bankMsgClient.Send(request1);
```

I'm guessing unfamiliar users would think "ok this returns a promise, so I should await", and that would not work. This, + the final `Promise.all()`, is not something I find really UX-friendly, so would prefer the callback version here.

But happy to define this as the go-to UX if it's well-documented.

Reply:
> so would prefer the callback version here.

Which is also easy to achieve without changing the code generator:

```ts
const signingClient = new SigningStargateClient(...);
const bankMsgClient = new MsgClientImpl(signingClient); // MsgClientImpl is imported from this file

// signingClient (or a wrapper around it) could hold an internal array of `Msg`s, and the following line would add one `Msg` to the array
bankMsgClient.Send(request1).then((response) => { console.log(response); }); // Everything will be well-typed.
// We could add more `Msg`s to the tx, by importing `Msg` clients from other modules
stakingMsgClient.Delegate(request2).then((response) => { console.log(response); });

// ...

signingClient.signAndBroadcast(...); // Signs and broadcasts tx with all internal `Msg`s.
```

Reply:
Those callbacks are much more annoying if you want to do anything with them (not console.log).

Let's say you want to use these results in a calculation. How do you get access to both in a callback? It is a nightmare (like async js coding 7 years ago). The Promise.all is one line to learn and then super easy to use the response.

Then again, I am a bit confused by your use-case here. If you show this hooked up to something, then maybe it is clearer

Reply:
Sounds good 👍.